examples/gateway-federation.js

```diff
@@ -0,0 +1,305 @@
+'use strict'
+
+const Fastify = require('fastify')
+const mercuriusGateway = require('@mercuriusjs/gateway')
+const mercuriusFederationPlugin = require('@mercuriusjs/federation')
+const redis = require('@fastify/redis')
+const fp = require('fastify-plugin')
+const cache = require('mercurius-cache')
+
+async function createPostService () {
+  const service = Fastify({ logger: true })
+
+  const posts = {
+    p1: {
+      pid: 'p1',
+      title: 'Post 1',
+      content: 'Content 1',
+      categoryId: 'c1'
+    },
+    p2: {
+      pid: 'p2',
+      title: 'Post 2',
+      content: 'Content 2',
+      categoryId: 'c2'
+    },
+    p3: {
+      pid: 'p3',
+      title: 'Post 3',
+      content: 'Content 3',
+      categoryId: 'c1'
+    },
+    p4: {
+      pid: 'p4',
+      title: 'Post 4',
+      content: 'Content 4',
+      categoryId: 'c1'
+    }
+  }
+
+  const schema = `
+    type Post @key(fields: "pid") {
+      pid: ID!
+      category: Category
+    }
+
+    type Query @extends {
+      posts: [Post]
+      topPosts(count: Int): [Post]
+    }
+
+    type Category @key(fields: "id") @extends {
+      id: ID! @external
+      topPosts(count: Int!): [Post]
+    }
+
+    input PostInput {
+      categoryId: ID!
+    }
+
+    type Mutation {
+      createPost(post: PostInput!): Post
+    }
+  `
+
+  const resolvers = {
+    Post: {
+      __resolveReference: (post, args, context, info) => {
+        return posts[post.pid]
+      },
+      category: (post, args, context, info) => {
+        context.app.log.info('Post.category')
+        return {
+          __typename: 'Category',
+          id: post.categoryId
+        }
+      }
+    },
+    Category: {
+      topPosts: (category, { count }, context, info) => {
+        context.app.log.info('Category.topPosts')
+        return Object.values(posts)
+          .filter((p) => p.categoryId === category.id)
+          .slice(0, count)
+      }
+    },
+    Query: {
+      posts: (root, args, context, info) => {
+        return Object.values(posts)
+      },
+      topPosts: (root, { count = 2 }, context, info) => {
+        context.app.log.info('Query.topPosts')
+        return Object.values(posts).slice(0, count)
+      }
+    },
+    Mutation: {
+      createPost: (root, { post }) => {
+        const pid = `p${Object.values(posts).length + 1}`
+
+        const result = {
+          pid,
+          ...post
+        }
+
+        posts[pid] = result
+        return result
+      }
+    }
+  }
+
+  service.register(mercuriusFederationPlugin, {
+    schema,
+    resolvers,
+    graphiql: true,
+    jit: 1
+  })
+
+  service.register(redis)
+
+  service.register(
+    fp(async (service) => {
```

Comment:
if `fp` needed?

Reply:
If I don't use `fp` I get the following error: 'Redis client is required'

Reply:
try awaiting all calls to .register

Reply:
It worked! I just updated the code.